contracts/GNTPaymentChannels.sol

```diff
@@ -39,157 +33,114 @@ contract GNTPaymentChannels is ReceivingContract {
         _;
     }
 
     // helpers: check channel status
 
-    function getDeposited(bytes32 _channel) external view returns (uint256) {
-        PaymentChannel storage ch = channels[_channel];
-        return ch.deposited;
-    }
-
-    function getWithdrawn(bytes32 _channel) external view returns (uint256) {
-        return channels[_channel].withdrawn;
+    function getDeposited(address owner, address receiver) external view returns (uint256) {
+        return _getChannel(owner, receiver).deposited;
     }
 
-    function getOwner(bytes32 _channel) external view returns (address) {
-        return channels[_channel].owner;
+    function getWithdrawn(address owner, address receiver) external view returns (uint256) {
+        return _getChannel(owner, receiver).withdrawn;
     }
 
-    function getReceiver(bytes32 _channel) external view returns (address) {
-        return channels[_channel].receiver;
+    function isLocked(address owner, address receiver) public view returns (bool) {
+        return _getChannel(owner, receiver).locked_until == 0;
     }
 
-    function isLocked(bytes32 _channel) public view returns (bool) {
-        return channels[_channel].locked_until == 0;
-    }
-
-    function isTimeLocked(bytes32 _channel) public view returns (bool) {
-        return channels[_channel].locked_until >= block.timestamp;
-    }
-
-    function isUnlocked(bytes32 _channel) public view returns (bool) {
-        return !isLocked(_channel) && !isTimeLocked(_channel);
+    function isTimeLocked(address owner, address receiver) public view returns (bool) {
+        return _getChannel(owner, receiver).locked_until >= block.timestamp;
     }
 
     function isValidSig(
-        bytes32 _ch,
-        uint _value,
+        address _owner,
+        address _receiver,
+        uint256 _value,
         uint8 _v,
         bytes32 _r,
         bytes32 _s
     )
         public
-        view
+        pure
         returns (bool)
     {
-        return (channels[_ch].owner) == (ecrecover(keccak256(_ch, _value), _v, _r, _s));
+        return _owner == ecrecover(keccak256(_owner, _receiver, _value), _v, _r, _s);
     }
 
     // functions that modify state
 
-    // FIXME: Channel needs to be created before it can be funded.
-    function createChannel(address _receiver) external {
-        bytes32 channel = keccak256(id++);
-        channels[channel] = PaymentChannel(msg.sender, _receiver, 0, 0, 0);
-        emit NewChannel(msg.sender, _receiver, channel); // event
-    }
-
     // Fund existing channel; can be done multiple times.
-    function onTokenReceived(address _from, uint _value, bytes _data) public onlyToken {
-        bytes32 channel;
+    function onTokenReceived(address _owner, uint256 _value, bytes _data) public onlyToken {
+        require(_data.length == 20);
+        address receiver;
         assembly {
-          channel := mload(add(_data, 32))
+          receiver := mload(add(_data, 20))
```

Comment:
So accidentally the above is in fact correct. The proper way (and I will change that) would be `receiver := div(mload(add(_data, 0x20)), 0x1000000000000000000000000)` though.

Why the above is correct?
The `_data` looks like this: `xxx...xxyyy...y` where `x` and `y` and single bytes and:
- there are 32 `x` which encode the length which is equal to 20
- there are 20 `y` which are the address
`_data` is a pointer to the first `x`
`add(_data, 20)` is the pointer to the 21st `x`
`mload(add(_data, 20))` then loads 32 bytes starting at the 21st `x`, so 12 `x` and all the `y`
since `receiver` is a variable of `address` type it ignores first 12 bytes leaving us with just the 20 `y` which are exactly the address.
So seems that it works, but I'm gonna change it because I wouldn't rely on the fact that those 12 extra `x` bytes are in fact always ignored (the new version ensures that those bytes will be zeroed).